# <center> Profitable App Profile for the Apple Store and Google Play Markets </center>

<center> by: Chlaudiah Julinar Soplero Lelywiary </center>

This project is to find the characteristics of mobile appplication that are prefreable in Apple Store and Google Play Store. As a data analyst who work in company that builds Android and iOS mobile apps, the job is to help the developers team to make a decision based on data-driven approach for decided what kind of apps that they should build. 

The company only build apps that are free to download and install and main source of revenue consites of in-app adds. This means our revenue for any given app is mostly influenced by the number of users who use our app. Our goal for this project is to analyze data to help our developers understand what kind of apps are likely to attract more users.

## Opening and Exploring the Data

As of September 2018, there were approximately 2 million iOS apps available on the App Store and 2.1 million Android apps on Google Play (source from: [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/))

Collectiong data over 4 million apps requires a significant amount of time and money, so try to analyze a sample of the data instead. To avoid spending resources on collectiong new data ourselves try to see any relevant existing data without cost. There are two datasets below seem suitable for the goals:

* [Google Apps Store Datasets](https://www.kaggle.com/lava18/google-play-store-apps/home) containing data about approximately **10.000** Android apps from Google Play that collected in August 2018.
* [Apple iOS Apps Store Datasets](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) containing data about approximately **7.000** iOS apps from the App Store that collected in July 2017.

Lets start by opening two data sets and then continue with exploring the data.

In [1]:
## OPEN TWO DATA SETS#

from csv import reader

#------- Google App Store Data sets------#
open_android = open('googleplaystore.csv',encoding='utf-8')
android_file = reader(open_android)
android_lists = list(android_file)
android_header = android_lists[0]
android_data = android_lists[1:]

#------- Apple iOS App Store Data sets------#
open_ios = open('AppleStore.csv',encoding='utf-8')
ios_file = reader(open_ios)
ios_lists = list(ios_file)
ios_header = ios_lists[0]
ios_data = ios_lists[1:]

In [2]:
# EXPLORE TWO DATA SETS #
def explore_data(lists, header=False):
    for row in lists[:3]:
        print(row,'\n')
        
    if header:
        print('Number of rows: ',len(lists))
        print('Number of columns: ',len(lists[0]))

In [3]:
print('Android Data Explore: ')
android_explore = explore_data(android_lists, header=True)

Android Data Explore: 
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

Number of rows:  10842
Number of columns:  13


As we can see from the result, for Android Dataset has **13 feature** and **10841 data**. The features that might be useful for our goals are:

| <center> Column Name </center> | <center> Description </center>                                |
| ----------- | ------------------------------------------- |
| App         | Application Name                            |
| Category    | Category the app belongs to                 |
| Rating      | Overall user rating of the app              |
| Reviews     | Number of user reviews                      |
| Installs    | Number of user download or install the app  |
| Type        | a paid app or free app                      |
| Genres      | Apart from the main category                |
| Price       | Price of the app                            |

In [4]:
print('iOS Data Explore: ')
ios_explore = explore_data(ios_lists,True)

iOS Data Explore: 
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

Number of rows:  7198
Number of columns:  16


For iOS data sets has **16 features** and **7197 data**. The features that might be useful are:

| Column Name       |  Description                             |
| -----------       | -----------------------------------------|
| id                | Application ID                           |
| track_name        | Application Name                         |
| price             | price amount                             |
| rating_count_tot  | Number of user rating count for all ver  |
| rating_count_ver  | Number of user rating count for cur ver  |
| prime_genre       | primary genre                            |
| currency          | Currency Type                            |

## Data Cleaning

### Detecting and Deleting Wrong Data

The Google Play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and [one of the discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for a certain row.



In [5]:
print('Wrong data: ',android_data[10472],'\n')
print('List of Features: ',android_header,'\n')
print('Example of right data: ',android_data[0])

Wrong data:  ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

List of Features:  ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

Example of right data:  ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


It is right from the discussion, that the data in index:**10472** (without header) is a wrong data because **Content Rating** feature has a blank value. Also, as we can see the **Rating** of that app reach **19**, and it is clearly off because the maximum rating for a Google Play App is a **5**.

So, need to delete the row of that wrong data from the Google Play data set.

In [6]:
del android_data[10472]
print(len(android_data))

10840


### Removing Duplicate Entries

#### Part One

After cleaning the wrong data, in this part will delete the duplicate entries in Google Play data set (look at the [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion)). Duplicate entries cases where an application appears more than once in a data set. For example, Instagram has four entries in Google Play data set:

In [7]:
for i in range(len(android_data)):
    name=android_data[i][0]
    if name=='Instagram':
        print(i,android_data[i])

2545 ['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
2604 ['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
2611 ['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
3909 ['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


We will count how many cases are duplicate entries in Google Play data set.

In [8]:
duplicate = []
unique=[]
for row in android_data:
    name=row[0]
    if name in unique:
        duplicate.append(name)
    else: unique.append(name)
print("Number of duplicate entries: ",len(duplicate))
print("Example of duplicate entries: ",duplicate[:5])

Number of duplicate entries:  1181
Example of duplicate entries:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


From the result, we can see that there are 1,181 cases for duplicate entries in Google Play datset. To handle this case, we don't want to count certain apps more than once when analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but there is still another way and better than this. 

If we back to Instagram's rows example, there are four difference in every row, and that is in index 3 (remember that list in Python start with 0). To find what column is it for index 3, then we can print it below:

In [9]:
for i in range(len(android_header)):
    print(i,android_header[i])

0 App
1 Category
2 Rating
3 Reviews
4 Size
5 Installs
6 Type
7 Price
8 Content Rating
9 Genres
10 Last Updated
11 Current Ver
12 Android Ver


So, the differences that occurs in every Instagram's row is number of reviews. The different numbers show the data was collected at different times. Using this information, the criterion for removing the duplicates entries can be build. Main important thing that we can get from this information is **the higher number of reviews, the more recent the data should be**.

#### Part Two

Instead of removing the duplicates randomly, just keep the row with the highest number of reviews and remove the other entries for any given app. To remove the duplicates, the steps are:
1. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app
2. Use the information stored in dictionary and create a new data set, which will have only entry per app (and for each app, only select the entry with the highest number of reviews)

##### Step 1

In this step, we build a dictionary to save the highest number of review from every app. 

In [10]:
reviews_max={}
for row in android_data:
    name = row[0]
    n_reviews=float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name]=n_reviews
    reviews_max[name]=n_reviews

In **Part One** we knew that there are **1,181 of duplicate entries** that we have. So, for make sure our dictionary doesnt have an app occurs more than once (of unique apps) it should be equal to the difference between the length of our data set and 1,181.

In [11]:
print("Expected Length: ",len(android_data)-1181)
print("Actual Length of Dictionary: ",len(reviews_max))
print("Example of the Highest number of review for Instagram: ",reviews_max["Instagram"])

Expected Length:  9659
Actual Length of Dictionary:  9659
Example of the Highest number of review for Instagram:  66509917.0


##### Step 2

Next, we need to make a new data set that stored only one entry per app with the highest number of reviews. In the code below:

- Create two empty lists, `android_clean` (will store new data set) and `already_dead` (will store app names)
- Do looping in Google Play data set without header (`android_data`) for each row:
    - Assign app name (in index 0)
    - Convert the type of number of reviews (in index 3) to `float`  and asssign it into a variable
    - Add the current row into `android_clean` and the app name into `already_dead`, **if**:
        - The number of reviews equal with the highest number of reviews from `reviews_max` dictionary for that app; and
        - The name of the app not already in `already_dead`. This part is important to handle a case when there are more than one entry has highest number of reviews for an app.


In [12]:
android_clean = []
already_dead=[]
for row in android_data:
    name = row[0]
    n_reviews = float(row[3])
    if (n_reviews == reviews_max[name]) and (name not in already_dead):
        android_clean.append(row)
        already_dead.append(name)

Now, just make sure that our new data set already same with the criterion using `explore_data` function that we already build and the number of rows is **9659**.

In [13]:
explore_data(android_clean,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  9659
Number of columns:  13


### Removing Non-English Apps

Remember that we use English for the apps we develop at our company, and we'd like to analyze only the apps that are directed toward an English-speaking audience. 

#### Part One

If we explore the data long enough, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets (iOS data sets and Android data sets that has been cleaned):

In [14]:
print(ios_data[813][1])
print(ios_data[6731][1])
print('-----------')
print(android_clean[4360][0])
print(android_clean[7939][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
-----------
中国語 AQリスニング
لعبة تقدر تربح DZ


We're not interested in keeping these apps. so we'll remove them. One way to go about this is to remove each app with a name containing symbol that is not commonly used in English text, which is usually includes letters from the English alphabet, number composed of digits from 0 to 9, punctuation marks (.,!,?,;), and other symbols (+,*,/)

All numbers corresponding to characters that commonly use in English text are in range 0-127, according to the ASCII. Based on this range, we can build a function that detects a character belongs to the set of common English characters or not. 

In [15]:
def isEnglish(strings):
    for char in strings:
        if ord(char) > 127:
            return False
            break
            
    return True

In [16]:
instagram = 'Instagram'
pps = '爱奇艺PPS -《欢乐颂2》电视剧热播'
docs='Docs To Go™ Free Office Suite'
instachat='Instachat 😜'
print(isEnglish(instagram))
print(isEnglish(pps))
print(isEnglish(docs))
print(isEnglish(instachat))

True
False
False
False


The `isEnglish()` function return right answer for `instagram` and `pps`, but return incorrect values for the next `docs` and `instachat`, which is the right values must be **True** because both are in English text. As you can see above, the last two strings has symbols that are not alphabet character. This results in the ASCII value for the symbols being more than 127, thereby returning the **False** value for both strings. Let see the number of ASCII from that symbols, below:

In [17]:
for char in docs:
    if ord(char)>127:
        print(char,ord(char))

for char in instachat:
    if ord(char)>127:
        print(char,ord(char))

™ 8482
😜 128540


From the result means that `isEnglish()` function need to be fix, to handle any kind of symbols that appeared in English text.

#### Part Two

In this part we will be fix the `isEnglish()` function, so we'll not going to lose useful data, since many English app will be incorrectly labeled as non-English. To minimize the impact of data loss, just remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range.

In [18]:
def isEnglish(strings):
    count=0
    for char in strings:
        if ord(char)>127:
            count+=1
    if count>3: return False
    return True

Lets try using `pps`, `docs`, and `instachat` strings to check the result of the new `isEnglish()` function (do not need to check with `instagram` because there are no symbol in that string).

In [19]:
pps = '爱奇艺PPS -《欢乐颂2》电视剧热播'
docs='Docs To Go™ Free Office Suite'
instachat='Instachat 😜'
print(isEnglish(instagram))
print(isEnglish(pps))
print(isEnglish(docs))
print(isEnglish(instachat))

True
False
True
True


As we can see, the function already return the right answer. Now, we will using `isEnglish()` to filter out non-English app from android data sets and iOS data sets.

In [20]:
android_english=[]
ios_english=[]

for row in android_clean:
    name=row[0]
    if isEnglish(name)==True:
        android_english.append(row)

for row in ios_data:
    name=row[1]
    if isEnglish(name)==True:
        ios_english.append(row)

In [21]:
explore_data(android_english, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows:  9614
Number of columns:  13


In [22]:
explore_data(ios_english, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1'] 

Number of rows:  6183
Number of columns:  16


After removing non-English app that did not match the criteria, now we just have **9614 data** from Android and **6183 data** from iOS.

### Isolating the Free Apps

As mentioned in introduction, the company just build apps that are free to download and installs, and our main source of revenue consists of in-app adds. But, both of data sets (Android and iOS) contain free and non-free apps. So, in this part we need to isolate only the free apps for our analysis. Isolating the free apps is last step in data cleaning process.

First, we need to know the number of column that has information about paid or free applications.

In [23]:
print(android_header)
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


From part of `Open and Exploring the Data`, for Android data sets it is **price** column in index 7, and for iOS data sets it is **price** column in index 4. Now, we can start to isolate the free apps below:

In [24]:
android_free=[]
ios_free=[]

for row in android_english:
    price = row[7]
    if  price == '0':
        android_free.append(row)
for row in ios_english:
    price = float(row[4])
    if price==0.0:
        ios_free.append(row)

In [25]:
print(len(android_free))
print(len(ios_free))

8864
3222


Now, we have **8864 free apps** for Android and **3222 free apps** for iOs.

## Data Analysis

After finish cleaning the data, now we start to analysis data sets to determine the kind of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:
    1. Build a minimal Android version of the app, and add it to Google Play
    2. If the app has a good response from users, we develop it further
    3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store

Because our end goal is to add the app on both Google Play and App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use gamification.

Before start our analysis, we can categorize the category of successfull app profiles based on Genre, and that are:
- Most Common Apps by Genre
- Most Popular Apps by Genre

### Most Common Apss by Genre

In this part, we will find the most common genre on the market based on the percentage of application with that genre in data set.

#### Part One

First, we need to know what feature that needed to find out the genre of apps in Android data sets and iOS data sets. From the part of `Open and Exploring the Data`, useful features and their descriptions have been determined and we can know that:
1. For Android data sets, features related to genre are `Category` and `Genres`. 
2. For iOS data sets, there is only one feature related to the genre namely `prime_genre`

Next, we need to find the index of that features.

In [26]:
for i in range(len(android_header)):
    if android_header[i]=='Category' or android_header[i]=='Genres':
        print(i,android_header[i])
for i in range(len(ios_header)):
    if ios_header[i]=='prime_genre':
        print(i,ios_header[i])

1 Category
9 Genres
11 prime_genre


So, for Android data sets we will use index `1` and `9` and for iOS data sets the index is `11`.

#### Part Two

Based on information above, we will build two functions we can use to analyze the frequency tables:
- One function to generate frequency tables that show percentages
- Another function we can use to display the percentages in a descending order

In [27]:
def freq_table(dataset,index):
    freq={}
    freq_percent={}
    total=0
    for row in dataset:
        total+=1
        genre=row[index]
        if genre in freq:
            freq[genre]+=1
        else: freq[genre]=1
    
    for genre in freq:
        freq_percent[genre]=(freq[genre]/total)*100
    
    return freq_percent

In [28]:
def display_table(dataset,index):
    table = freq_table(dataset,index)
    table_display=[]
    for genre in table:
        genre_percent_tuple = (table[genre],genre)
        table_display.append(genre_percent_tuple)
    
    table_sorted=sorted(table_display,reverse=True)
    for data in table_sorted:
        print(data[1],':',data[0])

##### Android Dat set Table

Based on `Category` in index `1`.

In [29]:
display_table(android_free,1)

FAMILY : 19.223826714801444
GAME : 9.510379061371841
TOOLS : 8.461191335740072
BUSINESS : 4.580324909747293
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.5424187725631766
SPORTS : 3.4183212996389893
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2490974729241873
HEALTH_AND_FITNESS : 3.068592057761733
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.782490974729242
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.128158844765343
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
ENTERTAINMENT : 0.8799638989169676
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 0

##### Analysis

From the result, the most common apps by Genre based on Category in Android Data sets is `Family` with the percentage around **19%**. Followed by `Game` (**9.5%**) and `Tools` (**8.5%**). Eventhough, Game is the second one with high percentage, but after Tools, other categories that follow are apps categories that are not designed for fun, instead have specific practical purpose such as business, lifestyle, productivity, finance, medical, etc.

Practical apps seems to have a good representation in Google Apps. It can be seen in the table percentage that the distribution of each category for Free English Apps not dominant to only one category, but almost evenly distributed in each category. 

Based on `Genres` in index `9`.

In [30]:
display_table(android_free,9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.580324909747293
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.5424187725631766
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2490974729241873
Action : 3.1024368231046933
Health & Fitness : 3.068592057761733
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.861462093862816
Video Players & Editors : 1.782490974729242
Casual : 1.7486462093862816
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.925090252707581

##### Analysis

The difference between the Genres and the Category columns is not clearly, but one thing we can notice is that the Genres column is much more specific because has more categories. Because we just want to look for the bigger picture at this moment, so working with the Category column is more recommended.

##### iOS Data Set Table

Based on `prime_category` in index `11`.

In [31]:
display_table(ios_free,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


##### Analysis

Among the Free English Apps in App Store, more than a half (**58.16%**) are `Games`. Followed by `Entertainment` around **8%**, and `Photo & Video` which are close to **5%**. Around **3%** there are apps with Education and Social Networking, followed by Shopping, Utilities, and Sports around **2%**. 

General information in App Store for Free English Apps is dominated by apps that are designed for fun, while apps with specific practical purpose are more rare. This is really different from Google Play Store, and also the distribution of Free English apps for each genre in App Store are not that good because it is to dominant in one of genre (Games). However, it doesn't mean that the most common apps also imply that they also have the greatest number of users. 

Up to this point, we found that App Store is dominated by Free English Apps designed for fun, meanwhile Google Play shows a more balanced distribution of both practical and for-fun apps. Now, we will find out about the kind of apps that have most users.

### Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Android data set, we can find this informations in the `Installs` column, but this information is missing for the iOS Store data set. As a workaround, we will take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.

Below, we calculate the average number of user ratings per app genre on the App Store.

In [32]:
for i in range(len(ios_header)):
    if ios_header[i]=='rating_count_tot':
        print(i,ios_header[i])

5 rating_count_tot


In [33]:
ios_freq=freq_table(ios_free,11)

In [34]:
genre_rating = []
for genre in ios_freq:
    total=0
    len_genre=0
    for row in ios_free:
        genre_app = row[11]
        if genre_app==genre:
            user_rating = float(row[5])
            total+=user_rating
            len_genre+=1
    avg_user_rating=total/len_genre
    genre_rating.append((avg_user_rating,genre))

genre_rating_sorted=sorted(genre_rating,reverse=True)
for data in genre_rating_sorted:
    print(data[1],':',data[0])

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


On average, `Navigation` apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together. 

In [35]:
for row in ios_free:
    if row[11]=='Navigation':
        print(row[1],':',row[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


This pattern also happened in Social Networking apps, where the average number is heavily influenced by a few giants like Facebook, Whatsapp, Pinterest, Skype, etc. Music apps's high average number influenced by a few big players like Spotify, Shazam, Pandora, etc.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to skewed by very few apps which have hundreds of thousands of user ratings, while other apps may struggle to get pas the 10,000 threshold. 

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [36]:
for row in ios_free:
    if row[11]=='Reference':
        print(row[1],':',row[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Other genres that seem popular include weather, book, food and drink, or finance. Lets take a time to talk about four of them:
- Wheater Apps --- people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.
- Book --- this seems overlap a bit with the app idea, but when people read a digital book they will spend a lot of time for that, which is they will open it more often. Also, because the main focus will be just a book, then we can add more features that focus in make our apps more fun. Nowadays, there are a lot of e-book.
- Food and drink --- examples here include Starbucks, McDonald's, Burger King's, etc. So, making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of the company
- Finance Apps --- these apps involve banking, paying bills, money transfer, etc. Building a financa app requires domain knowledge

From above, genres that seems to show potential is Book. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. The features like daily quotes from the book, an audio version of the book, quizzes about the book. Also, we could embed a dictionary within the app, so users don't need to exit our app to look up words in an external apps.

Now, let's analyze the Google Play Market.

### Most Popular Apps by Genre on Google Play

For Android data sets, we have data about the number of installs for the Google Play market. So, we should be able to get a clearer picture about genre popularity. 

In [37]:
for i in range(len(android_header)):
    if android_header[i]=='Installs':
        print(i,android_header[i])

5 Installs


In [38]:
display_table(android_free,5)

1,000,000+ : 15.749097472924186
100,000+ : 11.563628158844766
10,000,000+ : 10.503158844765343
10,000+ : 10.209837545126353
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


As we can see, the install numbers don't seem precise enough, because the most values are open-ended (100+, 1,000+, 5,000+). For instance, we don't know whethere an app with 100,000+ installs has 100,000 installs, 200,000, or 300,000. However, we don't need very precise data for our purposes, we only want to find out which app genres attract the most users, and we don't need perfect precision with respect to the number of users. 

We're going to leave the numbers as they are, which means that we will consider that an app with 100,000+ installs has 100,000 installs, and app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, we need to convert each install number from string to float. It means, we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error.

We will do that below, but remember that we will use the `Category` column in index `1`.

In [44]:
android_freq=freq_table(android_free,1)

In [46]:
category_install = []
for category in android_freq:
    total=0
    len_genre=0
    for row in android_free:
        category_app = row[1]
        if category_app==category:
            n_install = row[5]
            n_install = n_install.replace(',','')
            n_install = n_install.replace('+','')
            total+=float(n_install)
            len_genre+=1
    avg_n_install=total/len_genre
    category_install.append((avg_n_install,category))

category_install_sorted=sorted(category_install,reverse=True)
for data in category_install_sorted:
    print(data[1],':',data[0])

COMMUNICATION : 38326063.197916664
VIDEO_PLAYERS : 24790074.17721519
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16772838.591304347
TRAVEL_AND_LOCAL : 13984077.710144928
GAME : 12914435.883748516
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
ENTERTAINMENT : 9146923.076923076
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
FAMILY : 5180161.789906103
WEATHER : 5074486.197183099
SPORTS : 4274688.722772277
HEALTH_AND_FITNESS : 4167457.3602941176
MAPS_AND_NAVIGATION : 4056941.7741935486
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1768500.0
BUSINESS : 1704192.3399014778
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 513151.886792452

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook, Line, Google Chrome, etc), and a few others with over 100 and 500 million installs:

In [47]:
for row in android_free:
    if row[1]=='COMMUNICATION' and (row[5]=='1,000,000,000+'
                                   or row[5]=='500,000,000+'
                                   or row[5]=='100,000,000+'):
        print(row[0],':',row[5])

Messenger – Text and Video Chat for Free : 1,000,000,000+
Gmail : 1,000,000,000+
imo beta free calls and text : 100,000,000+
imo free video calls and chat : 500,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
WhatsApp Messenger : 1,000,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Hangouts : 1,000,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced ten times:

In [48]:
under_100_m=[]

for row in android_free:
    n_install = row[5]
    n_install = n_install.replace(',','')
    n_install = n_install.replace('+','')
    if row[1]=='COMMUNICATION' and float(n_install)<100000000:
        under_100_m.append(float(n_install))
sum(under_100_m)/len(under_100_m)

3593510.3486590036